src/shared/subprocess.cc

```diff
@@ -226,9 +215,27 @@ int SubProcess::kill(int signal) {
     return ret;
 }
 
-int SubProcess::terminate() {
+int SubProcess::hardkill() {
     auto ret = kill(SIGKILL);
-    wait();
+    wait(); // avoid zombies
+    return ret;
+}
+
+int SubProcess::terminate() {
+    // update state
+    poll();
+
+    // Graceful shutdown
+    auto ret = kill(SIGTERM);
+
+    for (int i = 0; i<20 && isRunning(); i++) {
+        usleep(100);
+        poll(); // update state after a while
+    }
+    if (isRunning()) {
+        // wait is already in hardkill
+        ret = hardkill(SIGKILL);
```

Comment:
hardkill has no arguments, compiled with C++ compiler this throws an error (but not in C compiler)

Reply:
Good point, and an oversight while renaming the routine and its usage... here's what our CIs are for ;)

Note that these sources use C++ constructs, so would not compile in C anyway.